Possible regression of pdb suppressing output

Hey guys, /cc @nicoddemus

Approx two years ago I submitted a patch for `--pdb` to stop suppressing output, seen here;
https://github.com/pytest-dev/pytest/issues/1223

I can see the PR was merged and the appropriate test is present in master;
https://github.com/pytest-dev/pytest/blob/5c6d7739bcd7a1ea33eecff37bd2c2271cf5309c/testing/test_pdb.py#L150

But when trying to use this myself, it doesn't seem to work, as seen below.

Any thoughts why this would pass unit tests, but wouldn't work in latest release?

```
user@c5595a22fd91:~/src/wtf$ cat test_all.py
def test_all():
    print('getrekt')
    assert False
```

```
user@c5595a22fd91:~/src/wtf$ pytest --pdb test_all.py
============================================================================================================ test session starts =============================================================================================================
platform linux -- Python 3.6.3, pytest-3.3.1, py-1.5.2, pluggy-0.6.0 -- /usr/bin/python3
cachedir: ../.cache
rootdir: /home/user/src, inifile: pytest.ini
collected 1 item
run-last-failure: run all (no recorded failures)

test_all.py::test_all FAILED                                                                                                                                                                                                           [100%]
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> traceback >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
Traceback (most recent call last):
  File "/home/user/src/wtf/test_all.py", line 3, in test_all
    assert False
AssertionError: assert False
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> entering PDB >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/user/src/wtf/test_all.py(3)test_all()
-> assert False
```

I started debugging this, and I don't yet know why the behavior broke, but I know why the test still passes. The test expects to find the text "getrekt" in the output:

`child.expect("getrekt")`

It indeed exists, just not as a result of the print statement, rather in the traceback of the assertion failure:
```
    def test_all():
        print("getrekt")
>       assert False
E       assert False
```

A possible solution is to update the print in the test to the following:

`print("get\\x20rekt")` (doubly escaped because it's a string inside a code string)

... and the regex to the following:

`child.expect("get rekt")`

Now the test does indeed fail, but passes if I add `-s` to the command line.

--- a/AUTHORS
+++ b/AUTHORS
@@ -29,6 +29,7 @@ Benjamin Peterson
 Bernard Pratz
 Bob Ippolito
 Brian Dorsey
+Brian Maissy
 Brian Okken
 Brianna Laugher
 Bruno Oliveira

--- a/_pytest/debugging.py
+++ b/_pytest/debugging.py
@@ -85,6 +85,17 @@ def _enter_pdb(node, excinfo, rep):
     # for not completely clear reasons.
     tw = node.config.pluginmanager.getplugin("terminalreporter")._tw
     tw.line()
+
+    captured_stdout = rep.capstdout
+    if len(captured_stdout) > 0:
+        tw.sep(">", "captured stdout")
+        tw.line(captured_stdout)
+
+    captured_stderr = rep.capstderr
+    if len(captured_stderr) > 0:
+        tw.sep(">", "captured stderr")
+        tw.line(captured_stderr)
+
     tw.sep(">", "traceback")
     rep.toterminal(tw)
     tw.sep(">", "entering PDB")

--- a/changelog/3052.bugfix
+++ b/changelog/3052.bugfix
@@ -1,0 +1,1 @@
+Added printing of captured stdout/stderr before entering pdb, and improved a test which was giving false negatives about output capturing.

--- a/testing/test_pdb.py
+++ b/testing/test_pdb.py
@@ -141,19 +141,50 @@ class TestPDB(object):
         child.sendeof()
         self.flush(child)
 
-    def test_pdb_interaction_capture(self, testdir):
+    def test_pdb_print_captured_stdout(self, testdir):
         p1 = testdir.makepyfile("""
             def test_1():
-                print("getrekt")
+                print("get\\x20rekt")
                 assert False
         """)
         child = testdir.spawn_pytest("--pdb %s" % p1)
-        child.expect("getrekt")
+        child.expect("captured stdout")
+        child.expect("get rekt")
         child.expect("(Pdb)")
         child.sendeof()
         rest = child.read().decode("utf8")
         assert "1 failed" in rest
-        assert "getrekt" not in rest
+        assert "get rekt" not in rest
+        self.flush(child)
+
+    def test_pdb_print_captured_stderr(self, testdir):
+        p1 = testdir.makepyfile("""
+            def test_1():
+                import sys
+                sys.stderr.write("get\\x20rekt")
+                assert False
+        """)
+        child = testdir.spawn_pytest("--pdb %s" % p1)
+        child.expect("captured stderr")
+        child.expect("get rekt")
+        child.expect("(Pdb)")
+        child.sendeof()
+        rest = child.read().decode("utf8")
+        assert "1 failed" in rest
+        assert "get rekt" not in rest
+        self.flush(child)
+
+    def test_pdb_dont_print_empty_captured_stdout_and_stderr(self, testdir):
+        p1 = testdir.makepyfile("""
+            def test_1():
+                assert False
+        """)
+        child = testdir.spawn_pytest("--pdb %s" % p1)
+        child.expect("(Pdb)")
+        output = child.before.decode("utf8")
+        child.sendeof()
+        assert "captured stdout" not in output
+        assert "captured stderr" not in output
         self.flush(child)
 
     def test_pdb_interaction_exception(self, testdir):

